File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int fun(int i, vector<int>& arr,vector<int>&dp) {
        if (i >= arr.size())
            return 0;
        if(dp[i]!=-1) return dp[i];
        int take = arr[i] + fun(i + 2, arr,dp);
        int notTake = fun(i + 1, arr,dp);

        return dp[i]=max(take, notTake);
    }

    int rob(vector<int>& nums) {

        if (nums.size() == 1)
            return nums[0];

        vector<int> temp1;
        vector<int> temp2;
        vector<int>dp1(nums.size(),-1);
        vector<int>dp2(nums.size(),-1);
        for (int i = 1; i < nums.size(); i++)
            temp1.push_back(nums[i]);

        for (int i = 0; i < nums.size() - 1; i++)
            temp2.push_back(nums[i]);

        return max(fun(0, temp1,dp1), fun(0, temp2,dp2));
    }
};

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def helper(self, i, nums, dp):
        if i >= len(nums):
            return 0

        if dp[i] != -1:
            return dp[i]

        take = nums[i] + self.helper(i + 2, nums, dp)
        notTake = self.helper(i + 1, nums, dp)

        dp[i] = max(take, notTake)

        return dp[i]

    def rob(self, nums: List[int]) -> int:
        n = len(nums)

        if n == 1:
            return nums[0]

        dp1 = [-1] * (n - 1)
        case1 = self.helper(0, nums[:n-1], dp1)

        dp2 = [-1] * (n - 1)
        case2 = self.helper(0, nums[1:], dp2)

        return max(case1, case2)